### ISSUE TITLE
BufferedAsyncFileStreamStandaloneConformanceTests.Read_EmptyStream_Nop_Success failing with PNSE

### ISSUE THREAD

**Issue by antonfirsov** [dotnet/runtime#44214]
### Test
`System.IO.Tests.BufferedAsyncFileStreamStandaloneConformanceTests.Read_EmptyStream_Nop_Success` 
(Multiple variants)

- Failing since ~30/10~ #44069
- On queues: `ubuntu.1804.amd64.open`, `windows.10.amd64.server19h1.es.open`, `windows.81.amd64.open`

https://dev.azure.com/dnceng/public/_build/results?buildId=874130&view=ms.vss-test-web.build-test-results-tab

#### Linux:
```
System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
```

#### Windows:
```
System.PlatformNotSupportedException : This API is specific to the way in which Windows handles asynchronous I/O, and is not supported on this platform.
```

#### Linux
<details>

```
   at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
   at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
   at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.IO.FileStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Stream.Read(Span`1 buffer)
   at System.IO.FileStream.Read(Span`1 buffer)
   at System.IO.Tests.StreamConformanceTests.ReadAsync(ReadWriteMode mode, Stream stream, Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at System.IO.Tests.StandaloneStreamConformanceTests.Read_EmptyStream_Nop_Success(ReadWriteMode mode)
--- End of stack trace from previous location ---
```

</details>

#### Windows
<details>

```
   at System.Threading.ThreadPoolBoundHandle.BindHandle(SafeHandle handle) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolBoundHandle.PlatformNotSupported.cs:line 24
   at System.IO.FileStream.Init(FileMode mode, FileShare share, String originalPath) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Windows.cs:line 91
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in /_/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs:line 252
   at System.IO.Tests.FileStreamStandaloneConformanceTests.CreateStream(Byte[] initialData, FileAccess access) in /_/src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.cs:line 23
   at System.IO.Tests.FileStreamStandaloneConformanceTests.CreateReadOnlyStreamCore(Byte[] initialData) in /_/src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.cs:line 26
   at System.IO.Tests.StandaloneStreamConformanceTests.CreateReadOnlyStream(Byte[] initialData) in /_/src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:line 653
   at System.IO.Tests.StandaloneStreamConformanceTests.Read_EmptyStream_Nop_Success(ReadWriteMode mode) in /_/src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs:line 833
--- End of stack trace from previous location ---
```

</details>

**Comment by stephentoub** [comment]
> Failing since 30/10

This test is new in master as of 17 hours ago.

> On queues: ubuntu.1804.amd64.open, windows.10.amd64.server19h1.es.open, windows.81.amd64.open

For reference, the key part isn't actually the queue, it's that it's mono wasm.

**Comment by antonfirsov** [comment]
> This test is new in master as of 17 hours ago.

Sorry for the misleading info, I've been seeing failures from #44069 in my Kusto query actually.

> For reference, the key part isn't actually the queue, it's that it's mono wasm.

Yeah I also suspected that, just wasn't able to find the relevant Kusto field.

**Comment by stephentoub** [comment]
Are you seeing the Windows failure you cited above in a PR against master, or just from my PR where I was writing the test?

**Comment by stephentoub** [comment]
And, actually, I'm not clear on why the cited test actually ran:
https://dev.azure.com/dnceng/public/_build/results?buildId=874130&view=ms.vss-test-web.build-test-results-tab
The base class of the test class has this on it:
https://github.com/dotnet/runtime/blob/5dcbc84a1b71b61addab65c414629f6f8f9bf41f/src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs#L21
@ViktorHofer, that should have prevented this from running, no?

**Comment by antonfirsov** [comment]
> Are you seeing the Windows failure you cited above in an active PR against master, or just from my PR where I was writing the test?

Windows failure seems to be specific to the PR, Ubuntu failures are happening in new PR-s.

I've also encountered a failure of `Read_PartiallySatisfied_RemainderOfBufferUntouched` in #44202, which *seems* to be related:
https://dev.azure.com/dnceng/public/_build/results?buildId=874355&view=ms.vss-test-web.build-test-results-tab
(`net6.0-Browser-Release-wasm-Mono_Release-wasmtestonbrowser-Ubuntu.1804.Amd64.Open`)

Same characteristics: started failing on Ubuntu today.

**Comment by stephentoub** [comment]
Trying a different mechanism for disabling it here: https://github.com/dotnet/runtime/pull/44218/files

**Comment by ViktorHofer** [comment]
> @ViktorHofer, that should have prevented this from running, no?

It should. The platform check is here: https://github.com/dotnet/arcade/blob/4873d157a8f34f8cc7e28b3f9938b32c642ef542/src/Microsoft.DotNet.XUnitExtensions/src/DiscovererHelpers.cs#L32.

@safern any idea why this didn't work? Both the conditionalclass attribute that stephen is using in his PR and the platformspecific attribute are using `ITraitDiscoverer`.

**Comment by safern** [comment]
> @safern any idea why this didn't work? Both the conditionalclass attribute that stephen is using in his PR and the platformspecific attribute are using ITraitDiscoverer.

I have no idea. Maybe something specific to xharness filtering with classes? They both look like should work just fine on classes. @akoeplinger do you see anything here?

**Comment by akoeplinger** [comment]
As I said in https://github.com/dotnet/runtime/issues/44226#issuecomment-721421101 this should work so no idea what's going on.

**Comment by stephentoub** [comment]
This just failed again here:
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-44248-merge-546b758713cf47cfb7/System.IO.FileSystem.Tests/console.33d6862b.log?sv=2019-07-07&se=2020-11-24T16%3A09%3A38Z&sr=c&sp=rl&sig=f6oDXMlCh2pt8hiX1DSGjqRKoj5e%2Fqi5M80R%2FsNRqoo%3D
and with employing a different suppression mechanism, ConditonalClass:
https://github.com/dotnet/runtime/blob/a627934ad271acf9def83d4ad5c4a5d1598c3106/src/libraries/Common/tests/Tests/System/IO/StreamConformanceTests.cs#L21
Something is definitely amiss... this test should not be running.

**Comment by ViktorHofer** [comment]
Just a random guess but try to attribute the actual classes with the attribute, not just the abstract base class.

**Comment by safern** [comment]
It could definitely be that it was applied to the abstract base class... I will debug a little bit later to understand what is going on.

**Comment by stephentoub** [comment]
I just tried this locally and all tests passed:
```C#
    [PlatformSpecific(~TestPlatforms.Windows)]
    public abstract class BaseClassWithPlatformSpecific
    {
        [Fact]
        public void Test1()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }

    public class DerivedClassWithBasePlatformSpecific : BaseClassWithPlatformSpecific
    {
        [Fact]
        public void Test2()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }

    [PlatformSpecific(~TestPlatforms.Windows)]
    public class PlatformSpecificWithoutBase
    {
        [Fact]
        public void Test3()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }

    [ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindows))]
    public abstract class BaseClassWithConditionalClass
    {
        [Fact]
        public void Test4()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }

    public class DerivedClassWithBaseConditionalClass : BaseClassWithConditionalClass
    {
        [Fact]
        public void Test5()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }

    [ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindows))]
    public class ConditionalClassWithoutBase
    {
        [Fact]
        public void Test6()
        {
            if (OperatingSystem.IsWindows()) throw new Exception("Shouldn't have run");
        }
    }
```
Maybe it's something specific with how the tests are executed for browser?

**Comment by safern** [comment]
I looked at the way XHarness for Browser does the test discovery and I don't see anything odd, it does everything the same as XUnit discoverer but in a single thread, but definitely something is going on there. I think we would need to debug it. I will open an issue on xharness.

**Comment by safern** [comment]
I opened: https://github.com/dotnet/xharness/issues/352
